### PR TITLE
fix(dom): preserve password input value state

### LINKED ISSUE #5795: Preserve filled state for redacted password inputs in DOM snapshots
## Problem

Browser Use now correctly avoids sending password values to the LLM, but the redaction path also removes the only observable indication that a password input is non-empty. A password field with a value and an empty password field are serialized identically. In form flows where the agent refreshes the DOM or verifies progress, the model can interpret an already-filled password as empty and type it repeatedly.

This is a follow-up to #5647/#5650: live values are now captured from `DOMSnapshot.captureSnapshot`, while sensitive values are intentionally excluded. The sensitive path needs to preserve field state without exposing the value.

## Reproduction

1. Navigate to a page containing two visible `input[type=password]` controls.
2. Set one control's live `.value` to a non-empty string and leave the other empty.
3. Capture the browser state and inspect the serialized DOM.

Both controls currently omit the live value and provide no state marker, so the LLM cannot distinguish them.

## Expected behavior

Keep the password, file, hidden, payment, and one-time-code values redacted, but expose a non-sensitive state marker such as `value-state=filled`, `value-state=empty`, or `value-state=unknown`.

The marker should be derived from the live snapshot value and must never include the value itself or its contents.

## Proposed fix

Store a boolean presence bit separately from the redacted value in `EnhancedSnapshotNode`, then emit the three-state marker only for password fields in the LLM-facing serializer. Add a regression test covering filled, empty, and cleared password inputs and asserting that the secret never appears in the serialized state.

### PR DESCRIPTION
**Pull request by Dante-dan**
## Why

Fixes #5795.

Password values are deliberately removed before they can reach logs or the LLM, but that also made filled and empty password controls indistinguishable. This caused agents to retype passwords after a DOM refresh even when the field was already populated.

## What changed

- Track a password input's live value as `filled`, `empty`, or `unknown` while processing `DOMSnapshot.captureSnapshot`, without retaining the value itself.
- Treat CDP's `-1` rare-string sentinel as an empty value and keep invalid references unknown.
- Emit `value-state` only for `input[type=password]`; other sensitive input types remain redacted without a new marker.
- Extend the existing live-input regression fixture with filled, empty, and cleared password controls and secret-leak assertions.

The resulting LLM view can distinguish states without receiving password content, for example:

```text
<input type=password value-state=filled />
<input type=password value-state=empty />
```

## Validation

- `./bin/lint.sh` — passed (ruff check/format, pre-commit hooks, pyright)
- `uv run pytest -q tests/ci/security/test_sensitive_data.py -k 'password_field or text_input_value_preserved' --tb=short` — 4 passed
- `uv run pytest -q tests/ci/test_dom_live_input_value.py --tb=short` — the test body passed with the filled/empty/cleared and non-leak assertions; the local process was interrupted after BrowserSession/http fixture teardown stopped completing
- `./bin/test.sh` — local run reached the repository's 11-failure stop threshold with 661 passed, 12 skipped, 3 failed, and 8 errors. The errors were dominated by concurrent Chrome processes exiting before CDP became available on the 16-worker run; the process was interrupted after it printed the failure summary and then stopped completing.

## Review follow-up

- Appended `input_value_state` after the existing `input_checked` field so `EnhancedSnapshotNode` preserves its pre-existing positional constructor order.
- `./bin/lint.sh` passes on the follow-up commit.
- A local rerun of `tests/ci/test_dom_live_input_value.py` could not start Chrome before CDP became available in the already exhausted local browser environment, then remained in teardown until interrupted. Current-head CI will rerun the browser regression.

---
## Summary by cubic
Fixes #5795. Redacted password values made filled and empty password controls indistinguishable, so agents retyped passwords after a DOM refresh. Password inputs now emit a `value-state` attribute (`filled`, `empty`, or `unknown`) in the LLM view while the value stays redacted.

- Treats CDP's `-1` rare-string sentinel as empty and invalid string references as `unknown`.
- Emits `value-state` only for `input[type=password]`; other sensitive input types stay redacted without a marker.
- Appends `input_value_state` after `input_checked` in `EnhancedSnapshotNode` to preserve the positional constructor order.

<sup>Written for commit 18f748ed53f24db90a4194f925f67d00aeeb3515. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/browser-use/browser-use/pull/5798?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: browser_use/dom/views.py [hunk 1]
```diff
@@ -355,6 +357,8 @@ class EnhancedSnapshotNode:
 	"""Stacking contexts from the layout tree"""
 	input_value: str | None = None
 	"""Live value of an <input> or <textarea> (DOMSnapshot inputValue/textValue), which the value attribute misses when JS, autofill, or a framework set it."""
+	input_value_state: InputValueState | None = None
```
[the remark this answers is not in the document]
Comment (Dante-dan) [reply browser_use/dom/views.py]: Good catch — this was a real compatibility issue even though all six in-repository construction sites use keyword arguments. I moved `input_value_state` after the existing `input_checked` field, preserving the pre-existing positional constructor order. `./bin/lint.sh` passes. A local rerun of the browser regression could not start Chrome before CDP became available in the already exhausted local browser environment and was interrupted during stuck teardown; current-head CI will rerun it.

**Comment by L4XB** [comment]
Not opening a competing PR — this is the same fix and it was here first. One measurement that may be worth having in the thread, because it turns your `-1` branch from a defensive guess into a confirmed requirement.

I had reached the opposite conclusion and was wrong. Probing `DOMSnapshot.captureSnapshot` on a page with one filled and three empty inputs, printing the raw `RareStringData` before anything interprets it:

```
inputValue  index= [6, 8, 10, 12]   value= [-1, 26, -1, -1]
textValue   index= []               value= []
```

So CDP does **not** omit empty inputs from the rare data, as I had assumed after reading a smaller page. It emits an entry for every input and uses `-1` as the empty-string sentinel — here three `-1`s (an untouched password, a password typed into and then cleared, and an empty text field) next to one real string index.

That matters for the shape of this fix. The obvious alternative — skip out-of-range string indices and treat "no entry for this node" as empty — lands on the right answer for the wrong reason: the entry *is* there, it is just skipped by the range check, and the correct result comes out of the fallback rather than out of the data. It would also flip to `unknown` for every empty field the moment that fallback is expressed slightly differently. Your explicit `string_index == -1 → 'empty'`, with genuinely invalid references kept `unknown`, reads the data as CDP actually sends it.

Two small things I would check while you are in there:

1. `password_value_state_by_index` seeds every password node with `'empty'` before the loop. Given the measurement above every password node does get an entry, so the seed is only reached if CDP ever stops emitting one — in which case `'empty'` is an assertion about a field nothing looked at. `'unknown'` would be the conservative seed, and costs nothing.
2. A field typed into and then cleared is the case that actually bit the reporter's workflow (the agent retypes into it). It shows up as `-1` above, so it is covered — worth a cell of its own in the fixture if there isn't one, since it is the only state that distinguishes "empty" from "never touched" in a user's mind.

I had this written before your PR appeared; the diff is at https://github.com/L4XB/browser-use/tree/fix/5795-password-value-state if any of the tests are useful. Nine mutations against it, all killed — including "the marker leaks the value" and "sensitive values are captured again", which are the two that matter for the redaction property. Happy for it to be reference material only.
